test/alternator/test_returnvalues.py

```diff
@@ -83,6 +83,28 @@ def test_put_item_returnvalues(test_table_s):
     with pytest.raises(ClientError, match='ValidationException'):
         test_table_s.put_item(Item={'p': p, 'a': 'hello'}, ReturnValues='none')
 
+# Testing feature which returns values only on delete condition failures.
+def test_put_item_returnvalues_on_condition_check_failure(test_table_s):
+    p = random_string()
+    test_table_s.put_item(Item={'p': p, 's': 'dog'})
+    # Succesful conditional put doesn't return values.
+    ret=test_table_s.put_item(Item={'p': p, 's': 'cat'},
+        ReturnValuesOnConditionCheckFailure='ALL_OLD',
+        ConditionExpression='s = :v1',
+        ExpressionAttributeValues={':v1' : 'dog'})
+    assert not 'Attributes' in ret
+    # Failed conditional delete returns old values.
+    got_exception = False
+    try:
+        test_table_s.put_item(Item={'p': p, 's': 'cow'},
+            ReturnValuesOnConditionCheckFailure='ALL_OLD',
+            ConditionExpression='s = :v1',
+            ExpressionAttributeValues={':v1' : 'dog'})
+    except test_table_s.meta.client.exceptions.ConditionalCheckFailedException as err:
+        assert err.response['Item'] == {'p': {'S': p}, 's': {'S': 'cat'}}
+        got_exception = True
+    assert got_exception
```

Comment:
nitpick: you can also use the more "idiomatic" pytest.raises, something like this (untested):
```python
with pytest.raises(ConditionalCheckFailedException) as err:
    test_table_s.put_item(...)
assert err.response['Item'] == ...
```
The raises idiom asserts that the exception must be thrown, and after the raises() check, you still have access to "err" and check that it's valid. So you don't need 3 lines dealing with "got_exception".

Reply:
thanks, your approach works, I just needed to add err.value prefix to get to response object